macros/provider.go

```diff
@@ -0,0 +1,155 @@
+package macros
+
+import (
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+	"github.com/prebid/prebid-server/exchange/entities"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	MacroKeyBidID       = "PBS-BIDID"
+	MacroKeyAppBundle   = "PBS-APPBUNDLE"
+	MacroKeyDomain      = "PBS-DOMAIN"
+	MacroKeyPubDomain   = "PBS-PUBDOMAIN"
+	MacroKeyPageURL     = "PBS-PAGEURL"
+	MacroKeyAccountID   = "PBS-ACCOUNTID"
+	MacroKeyLmtTracking = "PBS-LIMITADTRACKING"
+	MacroKeyConsent     = "PBS-GDPRCONSENT"
+	MacroKeyBidder      = "PBS-BIDDER"
+	MacroKeyIntegration = "PBS-INTEGRATION"
+	MacroKeyVastCRTID   = "PBS-VASTCRTID"
+	MacroKeyTimestamp   = "PBS-TIMESTAMP"
+	MacroKeyAuctionID   = "PBS-AUCTIONID"
+	MacroKeyChannel     = "PBS-CHANNEL"
+	MacroKeyEventType   = "PBS-EVENTTYPE"
+	MacroKeyVastEvent   = "PBS-VASTEVENT"
+)
+const (
+	CustomMacroPrefix = "PBS-MACRO-"
+)
+
+var (
+	bidLevelKeys = []string{MacroKeyBidID, MacroKeyBidder, MacroKeyVastEvent, MacroKeyEventType, MacroKeyVastCRTID}
+)
+
+type MacroContext struct {
+	Bid            *entities.PbsOrtbBid
+	Imp            *openrtb2.Imp
+	Seat           string
+	VastCreativeID string
+	VastEventType  string
+	EventElement   string
+}
+
+type macroProvider struct {
+	// macros map stores macros key values
+	macros map[string]string
+}
+
+// NewBuilder returns the instance of macro buidler
+func NewProvider(reqWrapper *openrtb_ext.RequestWrapper) *macroProvider {
+	macroProvider := &macroProvider{macros: map[string]string{}}
+	macroProvider.populateRequestMacros(reqWrapper)
+	return macroProvider
+}
+
+func (b *macroProvider) populateRequestMacros(reqWrapper *openrtb_ext.RequestWrapper) {
```

Comment:
Please also write unit tests for this function. I understand that this function is getting somewhat covered via other tests in `string_based_test.go` but I'd request you to consider adding separate unit test specifically for this function that exhaustively tests all the conditions.

Similarly please add test cases for `PopulateBidMacros` and `PopulateEventMacros`

Reply:
Add  unit tests for `PopulateBidMacros` and `PopulateEventMacros` and added single unit test for `populateRequestMacros` as cases separate test for each macro is covered under TestGetMacro. 
Is single unit test enough or should I write separate test for each macro?

Reply:
Added unit tests for `PopulateRequestMacros`,  `PopulateBidMacros` and `PopulateEventMacros`.